[TypeJoin] Implement Type::join for protocols and protocol compositions.

rudkx opened the pull request:
This is mostly complete, but we still need to add support for joins
between these two types and other types.

Commit:
[TypeJoin] Implement Type::join for protocols and protocol compositions.

This is mostly complete, but we still need to add support for joins
between these two types and other types.

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please test source compatibility

xedin reviewed (approved):
LGTM! Very nice!

slavapestov commented on lib/AST/TypeJoinMeet.cpp:
```diff
@@ -352,16 +345,130 @@ CanType TypeJoin::visitGenericFunctionType(CanType second) {
+  assert(First != second);
+
+  // FIXME: Handle other types here.
+  if (First->getKind() != TypeKind::Protocol &&
```
You should be using `isa<>` instead of `getKind()`, I think

rudkx replied:
Ah yes, I'll fix that up.

xedin replied:
Is that because isa<> strips sugar? I still can’t say when it would make sense to use getKind vs. isa<> with confidence... is getKind supposed to be used only by switch statements?

rudkx replied:
In general that's a good reason to use `is<>` and `getAs<>`, but in this case we're working with `CanType`s so that shouldn't make a difference (`*`). It's generally shorter and more idiomatic to use these templated functions.

Commit:
[TypeJoin] Replace getKind() with is<>(), and some other things.

Add a comment explaining that the distributive law isn't generally
true for lattices but should be for ours, and fix up the test case
to not use the real stdlib Strideable protocol in a couple places.

slavapestov commented on lib/AST/TypeJoinMeet.cpp:
```diff
@@ -352,16 +345,138 @@ CanType TypeJoin::visitGenericFunctionType(CanType second) {
+  assert(First != second);
+
+  // FIXME: Handle other types here.
+  if (!First->is<ProtocolType>() &&
```
This is First->isExistentialType()

rudkx commented on lib/AST/TypeJoinMeet.cpp:
```diff
@@ -352,16 +345,138 @@ CanType TypeJoin::visitGenericFunctionType(CanType second) {
+}
+
+// Return true if the first ProtocolDecl is a supertype of the second.
+static bool isSupertypeOf(ProtocolDecl *super, ProtocolDecl *sub) {
```
Close enough for my purposes, but the function I defined returns true if they are the same type. It shouldn't be calling them with the same type though, so I'll change it.

Commit:
[TypeJoin] Use isExistentialType() and inheritsFrom() as appropriate.

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please test source compatibility
